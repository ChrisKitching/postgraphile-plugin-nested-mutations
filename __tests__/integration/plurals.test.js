const { graphql } = require('graphql');
const { withSchema } = require('../helpers');

test(
  'plural when one-to-many, singular in reverse',
  withSchema({
    setup: `
      create table p.parent (
        id serial primary key,
        name text not null
      );
      
      create table p.child (
        id serial primary key,
        parent_id integer,
        name text not null,
        constraint child_parent_fkey foreign key (parent_id)
          references p.parent (id)
      );   
    `,
    test: async ({ schema, pgClient }) => {
      const query = `
        mutation {
          c1: createParent(
            input: {
              parent: {
                name: "test"
                childrenUsingId: {
                  create: [{
                    name: "test child"
                  }]
                }
              }
            }
          ) {
            parent {
              id
            }
          }

          c2: createChild(
            input: {
              child: {
                name: "child"
                parentToParentId: {
                  create: {
                    name: "child's parent"
                  }
                }
              }
            }
          ) {
            child {
              id
            }
          }
        }
      `;
      expect(schema).toMatchSnapshot();

      const result = await graphql(schema, query, null, { pgClient });
      expect(result).not.toHaveProperty('errors');
    },
  }),
);

test(
  'singular when one-to-one',
  withSchema({
    setup: `
      create table p.parent (
        id serial primary key,
        name text not null
      );
      
      create table p.child (
        parent_id serial primary key,
        name text not null,
        constraint child_parent_fkey foreign key (parent_id)
          references p.parent (id)
      );   
    `,
    test: async ({ schema, pgClient }) => {
      const query = `
        mutation {
          c1: createParent(
            input: {
              parent: {
                name: "test"
                childUsingId: {
                  create: {
                    name: "test child"
                  }
                }
              }
            }
          ) {
            parent {
              id
            }
          }

          c2: createChild(
            input: {
              child: {
                name: "child"
                parentToParentId: {
                  create: {
                    name: "child's parent"
                  }
                }
              }
            }
          ) {
            child {
              parentId
            }
          }
        }
      `;
      expect(schema).toMatchSnapshot();

      const result = await graphql(schema, query, null, { pgClient });
      expect(result).not.toHaveProperty('errors');
    },
  }),
);
