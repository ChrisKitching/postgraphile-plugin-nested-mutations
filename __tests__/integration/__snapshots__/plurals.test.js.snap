// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1:1 relationship does not allow multiple nested rows 1`] = `
GraphQLSchema {
  "__allowedLegacyNames": Array [],
  "__validationErrors": undefined,
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
  ],
  "_implementations": Object {
    "Node": Array [
      "Query",
      "Post",
      "PostImage",
    ],
  },
  "_mutationType": "Mutation",
  "_possibleTypeMap": undefined,
  "_queryType": "Query",
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "CreatePostImageInput": "CreatePostImageInput",
    "CreatePostImagePayload": "CreatePostImagePayload",
    "CreatePostInput": "CreatePostInput",
    "CreatePostPayload": "CreatePostPayload",
    "Cursor": "Cursor",
    "DeletePostByIdInput": "DeletePostByIdInput",
    "DeletePostImageByIdInput": "DeletePostImageByIdInput",
    "DeletePostImageByPostIdInput": "DeletePostImageByPostIdInput",
    "DeletePostImageInput": "DeletePostImageInput",
    "DeletePostImagePayload": "DeletePostImagePayload",
    "DeletePostInput": "DeletePostInput",
    "DeletePostPayload": "DeletePostPayload",
    "ID": "ID",
    "Int": "Int",
    "Mutation": "Mutation",
    "Node": "Node",
    "PageInfo": "PageInfo",
    "Post": "Post",
    "PostCondition": "PostCondition",
    "PostImage": "PostImage",
    "PostImageCondition": "PostImageCondition",
    "PostImageInput": "PostImageInput",
    "PostImageNodeIdConnect": "PostImageNodeIdConnect",
    "PostImageNodeIdUpdate": "PostImageNodeIdUpdate",
    "PostImagePatch": "PostImagePatch",
    "PostImagePostIdFkeyInput": "PostImagePostIdFkeyInput",
    "PostImagePostIdFkeyInverseInput": "PostImagePostIdFkeyInverseInput",
    "PostImagePostIdFkeyPostCreateInput": "PostImagePostIdFkeyPostCreateInput",
    "PostImagePostIdFkeyPostImageCreateInput": "PostImagePostIdFkeyPostImageCreateInput",
    "PostImagePostImagePkeyConnect": "PostImagePostImagePkeyConnect",
    "PostImagePostImagePkeyUpdate": "PostImagePostImagePkeyUpdate",
    "PostImagePostImagePostIdKeyConnect": "PostImagePostImagePostIdKeyConnect",
    "PostImagePostImagePostIdKeyUpdate": "PostImagePostImagePostIdKeyUpdate",
    "PostImagesConnection": "PostImagesConnection",
    "PostImagesEdge": "PostImagesEdge",
    "PostImagesOrderBy": "PostImagesOrderBy",
    "PostInput": "PostInput",
    "PostNodeIdConnect": "PostNodeIdConnect",
    "PostNodeIdUpdate": "PostNodeIdUpdate",
    "PostPatch": "PostPatch",
    "PostPostPkeyConnect": "PostPostPkeyConnect",
    "PostPostPkeyUpdate": "PostPostPkeyUpdate",
    "PostsConnection": "PostsConnection",
    "PostsEdge": "PostsEdge",
    "PostsOrderBy": "PostsOrderBy",
    "Query": "Query",
    "String": "String",
    "UpdatePostByIdInput": "UpdatePostByIdInput",
    "UpdatePostImageByIdInput": "UpdatePostImageByIdInput",
    "UpdatePostImageByPostIdInput": "UpdatePostImageByPostIdInput",
    "UpdatePostImageInput": "UpdatePostImageInput",
    "UpdatePostImagePayload": "UpdatePostImagePayload",
    "UpdatePostInput": "UpdatePostInput",
    "UpdatePostPayload": "UpdatePostPayload",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": undefined,
  "extensionASTNodes": undefined,
}
`;

exports[`1:1 relationship mutation fails when multiple operators are specified 1`] = `
GraphQLSchema {
  "__allowedLegacyNames": Array [],
  "__validationErrors": undefined,
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
  ],
  "_implementations": Object {
    "Node": Array [
      "Query",
      "Post",
      "PostImage",
    ],
  },
  "_mutationType": "Mutation",
  "_possibleTypeMap": undefined,
  "_queryType": "Query",
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "CreatePostImageInput": "CreatePostImageInput",
    "CreatePostImagePayload": "CreatePostImagePayload",
    "CreatePostInput": "CreatePostInput",
    "CreatePostPayload": "CreatePostPayload",
    "Cursor": "Cursor",
    "DeletePostByIdInput": "DeletePostByIdInput",
    "DeletePostImageByIdInput": "DeletePostImageByIdInput",
    "DeletePostImageByPostIdInput": "DeletePostImageByPostIdInput",
    "DeletePostImageInput": "DeletePostImageInput",
    "DeletePostImagePayload": "DeletePostImagePayload",
    "DeletePostInput": "DeletePostInput",
    "DeletePostPayload": "DeletePostPayload",
    "ID": "ID",
    "Int": "Int",
    "Mutation": "Mutation",
    "Node": "Node",
    "PageInfo": "PageInfo",
    "Post": "Post",
    "PostCondition": "PostCondition",
    "PostImage": "PostImage",
    "PostImageCondition": "PostImageCondition",
    "PostImageInput": "PostImageInput",
    "PostImageNodeIdConnect": "PostImageNodeIdConnect",
    "PostImageNodeIdUpdate": "PostImageNodeIdUpdate",
    "PostImagePatch": "PostImagePatch",
    "PostImagePostIdFkeyInput": "PostImagePostIdFkeyInput",
    "PostImagePostIdFkeyInverseInput": "PostImagePostIdFkeyInverseInput",
    "PostImagePostIdFkeyPostCreateInput": "PostImagePostIdFkeyPostCreateInput",
    "PostImagePostIdFkeyPostImageCreateInput": "PostImagePostIdFkeyPostImageCreateInput",
    "PostImagePostImagePkeyConnect": "PostImagePostImagePkeyConnect",
    "PostImagePostImagePkeyUpdate": "PostImagePostImagePkeyUpdate",
    "PostImagePostImagePostIdKeyConnect": "PostImagePostImagePostIdKeyConnect",
    "PostImagePostImagePostIdKeyUpdate": "PostImagePostImagePostIdKeyUpdate",
    "PostImagesConnection": "PostImagesConnection",
    "PostImagesEdge": "PostImagesEdge",
    "PostImagesOrderBy": "PostImagesOrderBy",
    "PostInput": "PostInput",
    "PostNodeIdConnect": "PostNodeIdConnect",
    "PostNodeIdUpdate": "PostNodeIdUpdate",
    "PostPatch": "PostPatch",
    "PostPostPkeyConnect": "PostPostPkeyConnect",
    "PostPostPkeyUpdate": "PostPostPkeyUpdate",
    "PostsConnection": "PostsConnection",
    "PostsEdge": "PostsEdge",
    "PostsOrderBy": "PostsOrderBy",
    "Query": "Query",
    "String": "String",
    "UpdatePostByIdInput": "UpdatePostByIdInput",
    "UpdatePostImageByIdInput": "UpdatePostImageByIdInput",
    "UpdatePostImageByPostIdInput": "UpdatePostImageByPostIdInput",
    "UpdatePostImageInput": "UpdatePostImageInput",
    "UpdatePostImagePayload": "UpdatePostImagePayload",
    "UpdatePostInput": "UpdatePostInput",
    "UpdatePostPayload": "UpdatePostPayload",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": undefined,
  "extensionASTNodes": undefined,
}
`;

exports[`plural when one-to-many, singular in reverse 1`] = `
GraphQLSchema {
  "__allowedLegacyNames": Array [],
  "__validationErrors": undefined,
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
  ],
  "_implementations": Object {
    "Node": Array [
      "Query",
      "Child",
      "Parent",
    ],
  },
  "_mutationType": "Mutation",
  "_possibleTypeMap": undefined,
  "_queryType": "Query",
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Child": "Child",
    "ChildChildPkeyConnect": "ChildChildPkeyConnect",
    "ChildChildPkeyUpdate": "ChildChildPkeyUpdate",
    "ChildCondition": "ChildCondition",
    "ChildInput": "ChildInput",
    "ChildNodeIdConnect": "ChildNodeIdConnect",
    "ChildNodeIdUpdate": "ChildNodeIdUpdate",
    "ChildParentFkeyChildCreateInput": "ChildParentFkeyChildCreateInput",
    "ChildParentFkeyInput": "ChildParentFkeyInput",
    "ChildParentFkeyInverseInput": "ChildParentFkeyInverseInput",
    "ChildParentFkeyParentCreateInput": "ChildParentFkeyParentCreateInput",
    "ChildPatch": "ChildPatch",
    "ChildrenConnection": "ChildrenConnection",
    "ChildrenEdge": "ChildrenEdge",
    "ChildrenOrderBy": "ChildrenOrderBy",
    "CreateChildInput": "CreateChildInput",
    "CreateChildPayload": "CreateChildPayload",
    "CreateParentInput": "CreateParentInput",
    "CreateParentPayload": "CreateParentPayload",
    "Cursor": "Cursor",
    "DeleteChildByIdInput": "DeleteChildByIdInput",
    "DeleteChildInput": "DeleteChildInput",
    "DeleteChildPayload": "DeleteChildPayload",
    "DeleteParentByIdInput": "DeleteParentByIdInput",
    "DeleteParentInput": "DeleteParentInput",
    "DeleteParentPayload": "DeleteParentPayload",
    "ID": "ID",
    "Int": "Int",
    "Mutation": "Mutation",
    "Node": "Node",
    "PageInfo": "PageInfo",
    "Parent": "Parent",
    "ParentCondition": "ParentCondition",
    "ParentInput": "ParentInput",
    "ParentNodeIdConnect": "ParentNodeIdConnect",
    "ParentNodeIdUpdate": "ParentNodeIdUpdate",
    "ParentParentPkeyConnect": "ParentParentPkeyConnect",
    "ParentParentPkeyUpdate": "ParentParentPkeyUpdate",
    "ParentPatch": "ParentPatch",
    "ParentsConnection": "ParentsConnection",
    "ParentsEdge": "ParentsEdge",
    "ParentsOrderBy": "ParentsOrderBy",
    "Query": "Query",
    "String": "String",
    "UpdateChildByIdInput": "UpdateChildByIdInput",
    "UpdateChildInput": "UpdateChildInput",
    "UpdateChildPayload": "UpdateChildPayload",
    "UpdateParentByIdInput": "UpdateParentByIdInput",
    "UpdateParentInput": "UpdateParentInput",
    "UpdateParentPayload": "UpdateParentPayload",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": undefined,
  "extensionASTNodes": undefined,
}
`;

exports[`singular when one-to-one 1`] = `
GraphQLSchema {
  "__allowedLegacyNames": Array [],
  "__validationErrors": undefined,
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
  ],
  "_implementations": Object {
    "Node": Array [
      "Query",
      "Child",
      "Parent",
    ],
  },
  "_mutationType": "Mutation",
  "_possibleTypeMap": undefined,
  "_queryType": "Query",
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Child": "Child",
    "ChildChildPkeyConnect": "ChildChildPkeyConnect",
    "ChildChildPkeyUpdate": "ChildChildPkeyUpdate",
    "ChildCondition": "ChildCondition",
    "ChildInput": "ChildInput",
    "ChildNodeIdConnect": "ChildNodeIdConnect",
    "ChildNodeIdUpdate": "ChildNodeIdUpdate",
    "ChildParentFkeyChildCreateInput": "ChildParentFkeyChildCreateInput",
    "ChildParentFkeyInput": "ChildParentFkeyInput",
    "ChildParentFkeyInverseInput": "ChildParentFkeyInverseInput",
    "ChildParentFkeyParentCreateInput": "ChildParentFkeyParentCreateInput",
    "ChildPatch": "ChildPatch",
    "ChildrenConnection": "ChildrenConnection",
    "ChildrenEdge": "ChildrenEdge",
    "ChildrenOrderBy": "ChildrenOrderBy",
    "CreateChildInput": "CreateChildInput",
    "CreateChildPayload": "CreateChildPayload",
    "CreateParentInput": "CreateParentInput",
    "CreateParentPayload": "CreateParentPayload",
    "Cursor": "Cursor",
    "DeleteChildByParentIdInput": "DeleteChildByParentIdInput",
    "DeleteChildInput": "DeleteChildInput",
    "DeleteChildPayload": "DeleteChildPayload",
    "DeleteParentByIdInput": "DeleteParentByIdInput",
    "DeleteParentInput": "DeleteParentInput",
    "DeleteParentPayload": "DeleteParentPayload",
    "ID": "ID",
    "Int": "Int",
    "Mutation": "Mutation",
    "Node": "Node",
    "PageInfo": "PageInfo",
    "Parent": "Parent",
    "ParentCondition": "ParentCondition",
    "ParentInput": "ParentInput",
    "ParentNodeIdConnect": "ParentNodeIdConnect",
    "ParentNodeIdUpdate": "ParentNodeIdUpdate",
    "ParentParentPkeyConnect": "ParentParentPkeyConnect",
    "ParentParentPkeyUpdate": "ParentParentPkeyUpdate",
    "ParentPatch": "ParentPatch",
    "ParentsConnection": "ParentsConnection",
    "ParentsEdge": "ParentsEdge",
    "ParentsOrderBy": "ParentsOrderBy",
    "Query": "Query",
    "String": "String",
    "UpdateChildByParentIdInput": "UpdateChildByParentIdInput",
    "UpdateChildInput": "UpdateChildInput",
    "UpdateChildPayload": "UpdateChildPayload",
    "UpdateParentByIdInput": "UpdateParentByIdInput",
    "UpdateParentInput": "UpdateParentInput",
    "UpdateParentPayload": "UpdateParentPayload",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": undefined,
  "extensionASTNodes": undefined,
}
`;
